// atom-ide-base/commons-ui/float-pane/ContentView.js
"use strict"
var __createBinding =
  (this && this.__createBinding) ||
  (Object.create
    ? function (o, m, k, k2) {
        if (k2 === undefined) k2 = k
        Object.defineProperty(o, k2, {
          enumerable: true,
          get: function () {
            return m[k]
          },
        })
      }
    : function (o, m, k, k2) {
        if (k2 === undefined) k2 = k
        o[k2] = m[k]
      })
var __setModuleDefault =
  (this && this.__setModuleDefault) ||
  (Object.create
    ? function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v })
      }
    : function (o, v) {
        o["default"] = v
      })
var __importStar =
  (this && this.__importStar) ||
  function (mod) {
    if (mod && mod.__esModule) return mod
    var result = {}
    if (mod != null)
      for (var k in mod)
        if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k)
    __setModuleDefault(result, mod)
    return result
  }
Object.defineProperty(exports, "__esModule", { value: true })
exports.ContentView = exports.DATATIP_ACTIONS = void 0
const React = __importStar(require("react"))
const HTMLView_1 = require("./HTMLView")
const SnippetView_1 = require("./SnippetView")
exports.DATATIP_ACTIONS = Object.freeze({
  PIN: "PIN",
  CLOSE: "CLOSE",
})
const IconsForAction = {
  [exports.DATATIP_ACTIONS.PIN]: "pin",
  [exports.DATATIP_ACTIONS.CLOSE]: "x",
}
async function getContent(datatip, editor, renderer) {
  var _a, _b
  let component = null
  let html = ""
  let snippet = ""
  if ("component" in datatip) {
    return React.createElement(datatip.component, null)
  } else if (datatip.markedStrings.length > 0) {
    const grammar = editor.getGrammar().scopeName.toLowerCase()
    snippet =
      (_a = await SnippetView_1.getSnippetHtml(
        datatip.markedStrings.filter((t) => t.type === "snippet").map((t) => t.value),
        grammar,
        renderer
      )) !== null && _a !== void 0
        ? _a
        : ""
    html =
      (_b = await HTMLView_1.getDocumentationHtml(
        datatip.markedStrings.filter((t) => t.type === "markdown").map((t) => t.value),
        grammar,
        renderer
      )) !== null && _b !== void 0
        ? _b
        : ""
    const htmlOrSnippet = snippet + html
    return React.createElement("div", null, htmlOrSnippet)
  } else {
    return null
  }
}
class ContentView extends React.Component {
  constructor() {
    super(...arguments)
    this.handleActionClick = (event) => {
      this.props.onActionClick()
    }
  }
  render() {
    const { className, action, actionTitle, datatip, onActionClick, ...props } = this.props
    let actionButton = null
    if (action != null && IconsForAction[action] != null) {
      const actionIcon = IconsForAction[action]
      actionButton = React.createElement("div", {
        className: `datatip-pin-button icon-${actionIcon}`,
        onClick: this.handleActionClick,
        title: actionTitle,
      })
    }
    const content = getContent(datatip, editor, renderer)
    return React.createElement(
      "div",
      Object.assign({ className: `${String(className)} datatip-container` }, props),
      React.createElement("div", { className: "datatip-content" }, content),
      actionButton
    )
  }
}
exports.ContentView = ContentView
