// atom-ide-base/commons-ui/float-pane/HTMLView.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDocumentationHtml = exports.HTMLView = void 0;
const dompurify_1 = __importDefault(require("dompurify"));
class HTMLView {
    constructor({ html, containerClassName, contentClassName, }) {
        this.rootElement = document.createElement("div");
        this.rootElement.className = containerClassName;
        this.rootElement.addEventListener("wheel", this.onMouseWheel, {
            passive: true,
        });
        if (html) {
            const innerHTML = dompurify_1.default.sanitize(html);
            this.rootElement.innerHTML = `
        <div className='${contentClassName}'>${innerHTML}</div>`;
        }
    }
    destroy() {
        this.rootElement.removeEventListener("wheel", this.onMouseWheel);
    }
    get element() {
        return this.rootElement;
    }
    onMouseWheel(evt) {
        evt.stopPropagation();
    }
}
exports.HTMLView = HTMLView;
async function getDocumentationHtml(markdownTexts, grammarName, renderer) {
    if (markdownTexts !== undefined && markdownTexts.length > 0) {
        const markdownText = markdownTexts.join("\r\n");
        return renderer.render(markdownText, grammarName);
    }
    return null;
}
exports.getDocumentationHtml = getDocumentationHtml;
