// atom-ide-base/commons-ui/float-pane/ReactView.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReactView = void 0;
const server_1 = require("react-dom/server");
const dompurify_1 = __importDefault(require("dompurify"));
class ReactView {
    constructor({ component, containerClassName, contentClassName, }) {
        this.containerClassName = containerClassName;
        this.contentClassName = contentClassName;
        if (component) {
            this.children = component();
        }
        this.render();
    }
    render() {
        this.rootElement = document.createElement("span");
        this.rootElement.classList.add(this.containerClassName);
        if (this.children) {
            this.childrenStatic = dompurify_1.default.sanitize(server_1.renderToStaticMarkup(this.children));
            this.rootElement.innerHTML = `
          <div className="${this.contentClassName}">${this.childrenStatic}</div>
      `;
        }
        return this.rootElement;
    }
    get element() {
        return this.rootElement;
    }
}
exports.ReactView = ReactView;
