// atom-ide-base/commons-ui/float-pane/ViewContainer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ViewContainer = void 0;
const etch_1 = __importDefault(require("etch"));
const HTMLView_1 = require("./HTMLView");
const SnippetView_1 = require("./SnippetView");
const ReactView_1 = require("./ReactView");
class ViewContainer {
    constructor(props, children) {
        this.props = props;
        this.children = children || [];
        this.updateChildren();
        etch_1.default.initialize(this);
    }
    render() {
        const glowClass = atom.config.get("atom-ide-datatip.glowOnHover") ? "datatip-glow" : "";
        const classes = `datatip-element ${glowClass}`;
        return etch_1.default.dom("div", { className: classes }, this.children);
    }
    update(props, children) {
        this.props = props;
        this.children = children || [];
        this.updateChildren();
        return etch_1.default.update(this);
    }
    async destroy() {
        await etch_1.default.destroy(this);
    }
    updateChildren() {
        if (this.props.component) {
            const { element, containerClassName, contentClassName } = this.props.component;
            this.children.push(etch_1.default.dom(ReactView_1.ReactView, { component: element, containerClassName: containerClassName, contentClassName: contentClassName }));
        }
        if (this.props.snippet) {
            const { element, containerClassName, contentClassName } = this.props.snippet;
            this.children.push(etch_1.default.dom(SnippetView_1.SnippetView, { snippet: element, containerClassName: containerClassName, contentClassName: contentClassName }));
        }
        if (this.props.html) {
            const { element, containerClassName, contentClassName } = this.props.html;
            this.children.push(etch_1.default.dom(HTMLView_1.HTMLView, { html: element, containerClassName: containerClassName, contentClassName: contentClassName }));
        }
    }
}
exports.ViewContainer = ViewContainer;
